contracts/performance/src/queries.rs

```diff
@@ -176,26 +234,63 @@ pub fn query_full_historical_performance_paged(
         .unwrap_or(retrieval_limits::NODE_HISTORICAL_PERFORMANCE_DEFAULT_LIMIT)
         .min(retrieval_limits::NODE_HISTORICAL_PERFORMANCE_MAX_LIMIT) as usize;
 
-    let start = start_after.map(Bound::exclusive);
+    // because results are sorted first by epoch_id, then by node_id, start at the next node_id
+    // (from the lexicographically first measurement_kind, which is empty string)
+    let start = start_after.map(|(epoch, node)| Bound::exclusive((epoch, node + 1, String::new())));
 
-    let performance = NYM_PERFORMANCE_CONTRACT_STORAGE
+    // storage keeps (epoch_id, node_id, measurement_kind) tuples,
+    // but the API needs results aggregated by (epoch_id, node_id) pairs
+    // so we need an intermediary struct that collects all measurements
+    // per (epoch_id, node_id) pair (and calculates a performance)
+    let mut res_per_epoch_and_node: HashMap<(EpochId, NodeId), HashMap<MeasurementKind, Percent>> =
+        HashMap::new();
+    NYM_PERFORMANCE_CONTRACT_STORAGE
         .performance_results
         .results
         .range(deps.storage, start, None, Order::Ascending)
-        .take(limit)
+        // we can't cut a pagination limit here becasue we don't want to
```

Comment:
why not. just change the query parameters and let the caller resume from where it's meant to resume from